Client credentials store.

calavera opened the pull request:
This change implements communication with an external credentials store,
ala git-credential-helper. The client falls back the plain text store,
what we're currently using, if there is no remote store configured.

This implementation shells out to helper program when a credential store is
configured. Those programs can be implemented with any language as long as they
follow the convention to pass arguments and information.

There is an implementation for the OS X keychain in https://github.com/docker/docker-credential-helpers. That package also provides basic structure to create other helpers.

![docker-osxkeychain](https://cloud.githubusercontent.com/assets/1050/12891685/79986130-ce3d-11e5-8225-5c6142ec31e1.gif)

Signed-off-by: David Calavera david.calavera@gmail.com

calavera committed:
Client credentials store.

This change implements communication with an external credentials store,
ala git-credential-helper. The client falls back the plain text store,
what we're currently using, if there is no remote store configured.

It shells out to helper program when a credential store is
configured. Those programs can be implemented with any language as long as they
follow the convention to pass arguments and information.

There is an implementation for the OS X keychain in https://github.com/calavera/docker-credential-helpers.
That package also provides basic structure to create other helpers.

Signed-off-by: David Calavera <david.calavera@gmail.com>

GordonTheTurtle commented:
Job: Docker-PRs FAILED
Here are what we parsed from the logs at https://jenkins.dockerproject.org/job/Docker-PRs/23550/consoleText as the failures:

~~~console

---
mt)
Congratulations!  All Go source files are properly formatted.

---> Making bundle: validate-lint (in bundles/1.11.0-dev/validate-lint)
Congratulations!  All Go source files have been linted.

---> Making bundle: validate-pkg (in bundles/1.11.0-dev/validate-pkg)
Congratulations! "./pkg/..." is safely isolated from internal code.

---> Making bundle: validate-test (in bundles/1.11.0-dev/validate-test)
Congratulations! No testing.T found.

---> Making bundle: validate-toml (in bundles/1.11.0-dev/validate-toml)
Congratulations!  All toml source files changed here have valid syntax.

---> Making bundle: validate-vet (in bundles/1.11.0-dev/validate-vet)
Congratulations!  All Go source files have been vetted.

---> Making bundle: binary (in bundles/1.11.0-dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
# github.com/docker/docker/cliconfig
cliconfig/credentials.go:69: not enough arguments to return
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs] $ /bin/sh -xe /tmp/hudson5401678874764349502.sh
+ set +x


---
Now starting POST-BUILD steps
---

Error response from daemon: no such id: docker-pr23550
Error: failed to remove containers: [docker-pr23550]
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-userns FAILED
Here are what we parsed from the logs at https://jenkins.dockerproject.org/job/Docker-PRs-userns/5988/consoleText as the failures:

~~~console

---
i/toml/cmd/tomlv 	&& rm -rf "$GOPATH"
 ---> Using cache
 ---> 499b70280da8
Step 46 : ENV RSRC_COMMIT ba14da1f827188454a4591717fff29999010887f
 ---> Using cache
 ---> 4ca6a3b43544
Step 47 : RUN set -x 	&& export GOPATH="$(mktemp -d)" 	&& git clone https://github.com/akavel/rsrc.git "$GOPATH/src/github.com/akavel/rsrc" 	&& (cd "$GOPATH/src/github.com/akavel/rsrc" && git checkout -q "$RSRC_COMMIT") 	&& go build -v -o /usr/local/bin/rsrc github.com/akavel/rsrc 	&& rm -rf "$GOPATH"
 ---> Using cache
 ---> 4dfd5bccc06d
Step 48 : ENTRYPOINT hack/dind
 ---> Using cache
 ---> 76079e7fffe9
Step 49 : COPY . /go/src/github.com/docker/docker
 ---> 16b4e1973702
Removing intermediate container 0878d39d8c2a
Successfully built 16b4e1973702

---> Making bundle: binary (in bundles/1.11.0-dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
# github.com/docker/docker/cliconfig
cliconfig/credentials.go:69: not enough arguments to return
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-userns] $ /bin/sh -xe /tmp/hudson783302646154928808.sh
+ set +x


---
Now starting POST-BUILD steps
---

Error response from daemon: no such id: docker-pr-userns5988
Error: failed to remove containers: [docker-pr-userns5988]
ERROR: Build step failed with exception
java.lang.IllegalStateException: Invalid object ID 62 iota=63
	at hudson.remoting.ExportTable.diagnoseInvalidId(ExportTable.java:386)
	at hudson.remoting.ExportTable.get(ExportTable.java:330)
	at hudson.remoting.Channel.getExportedObject(Channel.java:633)
	at hudson.remoting.RemoteInvocationHandler$RPCRequest.perform(RemoteInvocationHandler.java:599)
	at hudson.remoting.RemoteInvocationHandler$RPCRequest.call(RemoteInvocationHandler.java:583)
	at hudson.remoting.RemoteInvocationHandler$RPCRequest.call(RemoteInvocationHandler.java:542)
	at hudson.remoting.UserRequest.perform(UserRequest.java:120)
	at hudson.remoting.UserRequest.perform(UserRequest.java:48)
	at hudson.remoting.Request$2.run(Request.java:326)
	at hud
---

---
rService.java:68)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
	at ......remote call to ubuntu (aufs) (i-fd9b293a)(Native Method)
	at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1416)
	at hudson.remoting.UserResponse.retrieve(UserRequest.java:220)
	at hudson.remoting.Channel.call(Channel.java:781)
	at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:250)
	at com.sun.proxy.$Proxy86.join(Unknown Source)
	at hudson.Launcher$RemoteLauncher$ProcImpl.join(Launcher.java:991)
	at hudson.tasks.CommandInterpreter.join(CommandInterpreter.java:135)
	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:95)
	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:64)
	at org.jenkinsci.plugins.postbuildscript.PostBuildScript.processBuildSteps(PostBuildScript.java:204)
	at org.jenkinsci.plugins.postbuildscript.PostBuildScript.processScripts(PostBuildScript.java:143)
	at org.jenkinsci.plugins.postbuildscript.PostBuildScript._perform(PostBuildScript.java:105)
	at org.jenkinsci.plugins.postbuildscript.PostBuildScript.perform(PostBuildScript.java:85)
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:782)
	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:723)
	at hudson.model.Build$BuildExecution.post2(Build.java:185)
	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:668)
	at hudson.model.Run.execute(Run.java:1763)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:410)
Caused by: java.lang.Exception: Object was recently deallocated
    #62 (ref.0) : object=null 
---

---
t.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
	at ......remote call to ubuntu (aufs) (i-fd9b293a)(Native Method)
	at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1416)
	at hudson.remoting.UserResponse.retrieve(UserRequest.java:220)
	at hudson.remoting.Channel.call(Channel.java:781)
	at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:250)
	at com.sun.proxy.$Proxy86.join(Unknown Source)
	at hudson.Launcher$RemoteLauncher$ProcImpl.join(Launcher.java:991)
	at hudson.tasks.CommandInterpreter.join(CommandInterpreter.java:135)
	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:95)
	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:64)
	at org.jenkinsci.plugins.postbuildscript.PostBuildScript.processBuildSteps(PostBuildScript.java:204)
	at org.jenkinsci.plugins.postbuildscript.PostBuildScript.processScripts(PostBuildScript.java:143)
	at org.jenkinsci.plugins.postbuildscript.PostBuildScript._perform(PostBuildScript.java:105)
	at org.jenkinsci.plugins.postbuildscript.PostBuildScript.perform(PostBuildScript.java:85)
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:782)
	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:723)
	at hudson.model.Build$BuildExecution.post2(Build.java:185)
	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:668)
	at hudson.model.Run.execute(Run.java:1763)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:410)
Caused by: java.lang.Exception: Object was recently deallocated
    #62 (ref.0) : object=null type=hudson.Launcher$RemoteLaunchCallable$1 interfaces=[hudson.Launcher$RemoteProcess]
      Create
---

---
cutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
	at ......remote call to ubuntu (aufs) (i-fd9b293a)(Native Method)
	at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1416)
	at hudson.remoting.UserResponse.retrieve(UserRequest.java:220)
	at hudson.remoting.Channel.call(Channel.java:781)
	at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:250)
	at com.sun.proxy.$Proxy86.join(Unknown Source)
	at hudson.Launcher$RemoteLauncher$ProcImpl.join(Launcher.java:991)
	at hudson.tasks.CommandInterpreter.join(CommandInterpreter.java:135)
	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:95)
	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:64)
	at org.jenkinsci.plugins.postbuildscript.PostBuildScript.processBuildSteps(PostBuildScript.java:204)
	at org.jenkinsci.plugins.postbuildscript.PostBuildScript.processScripts(PostBuildScript.java:143)
	at org.jenkinsci.plugins.postbuildscript.PostBuildScript._perform(PostBuildScript.java:105)
	at org.jenkinsci.plugins.postbuildscript.PostBuildScript.perform(PostBuildScript.java:85)
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:782)
	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:723)
	at hudson.model.Build$BuildExecution.post2(Build.java:185)
	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:668)
	at hudson.model.Run.execute(Run.java:1763)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:410)
Caused by: java.lang.Exception: Object was recently deallocated
    #62 (ref.0) : object=null type=hudson.Launcher$RemoteLaunchCallable$1 interfaces=[hudson.Launcher$RemoteProcess]
      Created at Mon Feb 08 16:53:18 UTC 2016
    	at hudson.remoting.ExportTable$Entry.<init>(ExportTabl
---

---
.....remote call to ubuntu (aufs) (i-fd9b293a)(Native Method)
	at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1416)
	at hudson.remoting.UserResponse.retrieve(UserRequest.java:220)
	at hudson.remoting.Channel.call(Channel.java:781)
	at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:250)
	at com.sun.proxy.$Proxy86.join(Unknown Source)
	at hudson.Launcher$RemoteLauncher$ProcImpl.join(Launcher.java:991)
	at hudson.tasks.CommandInterpreter.join(CommandInterpreter.java:135)
	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:95)
	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:64)
	at org.jenkinsci.plugins.postbuildscript.PostBuildScript.processBuildSteps(PostBuildScript.java:204)
	at org.jenkinsci.plugins.postbuildscript.PostBuildScript.processScripts(PostBuildScript.java:143)
	at org.jenkinsci.plugins.postbuildscript.PostBuildScript._perform(PostBuildScript.java:105)
	at org.jenkinsci.plugins.postbuildscript.PostBuildScript.perform(PostBuildScript.java:85)
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:782)
	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:723)
	at hudson.model.Build$BuildExecution.post2(Build.java:185)
	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:668)
	at hudson.model.Run.execute(Run.java:1763)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:410)
Caused by: java.lang.Exception: Object was recently deallocated
    #62 (ref.0) : object=null type=hudson.Launcher$RemoteLaunchCallable$1 interfaces=[hudson.Launcher$RemoteProcess]
      Created at Mon Feb 08 16:53:18 UTC 2016
    	at hudson.remoting.ExportTable$Entry.<init>(ExportTable.java:99)
    	at hudson.remoting.ExportTable.export(ExportTable.java:305)
    	at hudson.
---

~~~

jfrazelle commented:
+1 I think we should use the native keychain and allow the operating system apply its security to the objects it controls

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 FAILED
Here are what we parsed from the logs at https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/742/consoleText as the failures:

~~~console

---
nary; \
	ec=$?; \
	if [ $ec -eq 0 ]; then \
		robocopy /c/go/src/github.com/docker/docker/bundles/$(cat VERSION)/binary /c/target/binary; \
	fi; \
	exit $ec'
C:\Windows\system32\docker.exe: Error response from daemon: Cannot start container 5a282eb32a5aa633766b17ee3e9131115dc6422dab71216a1e87dee45df6d130: HCSShim::StartComputeSystem- Win32 API call returned error r1=0x800710d8 err=The object identifier does not represent a valid object.id=5a282eb32a5aa633766b17ee3e9131115dc6422dab71216a1e87dee45df6d130.
+ ec=125
+ set +x

----------------------------------
ERROR: Failed to build test binary
----------------------------------



-----------------------------------------------
ERROR: Failed with exitcode 125 at Mon Feb  8 16:59:06 CUT 2016.
-----------------------------------------------


INFO: Tidying up at end of run
INFO: Nuking /d/CI
INFO: Zapped successfully
INFO: End of cleanup
INFO: Ended at Mon Feb  8 16:59:07 CUT 2016 (1m 1s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe D:\temp\hudson7390039133466014546.sh
+ set +e
+ set +x
INFO: End of cleanup
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-gccgo FAILED
Here are what we parsed from the logs at https://jenkins.dockerproject.org/job/Docker-PRs-gccgo/1669/consoleText as the failures:

~~~console

---
 13 : WORKDIR /go/src/github.com/docker/docker
 ---> Running in 049fbcaedc09
 ---> 7311057e01e7
Removing intermediate container 049fbcaedc09
Step 14 : ENV DOCKER_BUILDTAGS apparmor seccomp selinux
 ---> Running in ca1d0cc5b602
 ---> f323dc6d9a8a
Removing intermediate container ca1d0cc5b602
Step 15 : ENTRYPOINT hack/dind
 ---> Running in 07383c9b213b
 ---> 03d967483b2d
Removing intermediate container 07383c9b213b
Step 16 : COPY . /go/src/github.com/docker/docker
 ---> 4a8168e9efc2
Removing intermediate container 52667121b32c
Successfully built 4a8168e9efc2
# WARNING! I don't seem to be running in a Docker container.
# The result of this command might be an incorrect build, and will not be
# officially supported.
#
# Try this instead: make all
#

---> Making bundle: gccgo (in bundles/1.11.0-dev/gccgo)
# github.com/docker/docker/cliconfig
cliconfig/credentials.go:69:3: error: not enough arguments to return
   return err
   ^
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-gccgo] $ /bin/sh -xe /tmp/hudson5736975324661069332.sh
+ set +x


---
Now starting POST-BUILD steps
---

Error response from daemon: no such id: docker-pr-gccgo1669
Error: failed to remove containers: [docker-pr-gccgo1669]
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

thaJeztah commented:
nice! +1

stevvooe commented on vendor/src/github.com/hashicorp/go-plugin/LICENSE:
```diff
@@ -0,0 +1,353 @@
+Mozilla Public License, version 2.0
```
Is MPL compatible with our licensing and distribution?

calavera replied:
yes.

stevvooe replied:
Really? MPL is copyleft, whereas Apache is doesn't have such distribution restrictions. I don't see MPL anywhere on [here](https://www.docker.com/components-licenses). There is some GPLv3 but that specifically has a static linking exception.

IANAL, but this requires more thought. For example, this may have implications for proprietary authorization plugins.

calavera replied:
https://opensource.com/law/11/9/mpl-20-copyleft-and-license-compatibility

```
A good example of these subtle incompatibilities is between MPL 1.1 and Apache 2.0. In certain circumstances, a patent litigant could trigger Apache 2.0's patent peace clause without triggering MPL 1.1's similar (but slightly different) patent peace clause. This made the two licenses incompatible in the sense used here. Aligning these triggering conditions is what allowed us to be more comfortable that **Apache 2.0 and MPL 2.0 are compatible*. In other words, complying with MPL 2.0's patent requirements will allow someone to be comfortable that they've also complied with the patent requirements of any Apache-licensed code contained within the MPL-licensed project.
```

stevvooe replied:
I am not asking about license compatibility. I am asking about whether or not MPL is compatible with the goals of a login system. Are we comfortable with forcing implementors of login plugins to have to link against code under a copyleft license?

This is going to require more than a one word answer and a link to blog.

stevvooe commented on cliconfig/credentials/store.go:
```diff
@@ -0,0 +1,10 @@
+// Store is an interface that credentials providers must implement
+type Store interface {
+	Add(types.AuthConfig) error
+	Get(serverURL string) (string, string, error)
```
Is this assuming we want to index these by the `ServerAddress` field?

calavera replied:
yep, it only uses the server address for now, it could be extended to use serverAddress && username.

stevvooe commented on cliconfig/credentials.go:
```diff
@@ -0,0 +1,101 @@
+)
+
+// GetCredentials loads the user credentials for a given registry server.
+func (c *ConfigFile) GetCredentials(serverAddress string) (types.AuthConfig, error) {
```
Does it make sense architecturally to have the configuration file take on the role of saving credentials?

stevvooe commented:
This is pretty awesome.

I disagree with adopting yet-another-plugin-framework, but this is a fine start. Could we just define a cli API, like `credentials add|get|remove <serverurl>`?

calavera commented:
> I disagree with adopting yet-another-plugin-framework, but this is a fine star.

I'm not sold on that part either.

> Could we just define a cli API, like credentials add|get|remove <serverurl>? This would allow non-go plugins.

Yes, that's what git does. I wanted to have a working version before settling down in one protocol.

calavera commented:
What I like about this implementation is that it doesn't require any STDOUT parsing, like a pure command line would require. I don't think it would be much complicated, but it's definitely less "graceful".

vdemeester commented:
waw nice !! :+1: on this too :wink:

stevvooe commented:
> What I like about this implementation is that it doesn't require any STDOUT parsing, like a pure command line would require.

Piping stdout into a json unmarshaler seems straightforward and can be done with a large number of languages. The framing and protocol of go-plugin would have to implemented in each language. I don't think there is anything "graceful" about using go-plugin here.

calavera commented:
Okay, let's go with a shelling out implementation, it should not be very hard to modify and I agree that it's more open to other languages.

stevvooe commented:
@calavera :+1:

cpuguy83 commented:
:heart_eyes:

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/799/console):

~~~console

---
9d3f2d.
+ ec=125
+ set +x

----------------------------------
ERROR: Failed to build test binary
----------------------------------



-----------------------------------------------
ERROR: Failed with exitcode 125 at Tue Feb  9 21:28:28 CUT 2016.
-----------------------------------------------


INFO: Tidying up at end of run
INFO: Nuking /d/CI
INFO: Zapped successfully
INFO: End of cleanup
INFO: Ended at Tue Feb  9 21:28:29 CUT 2016 (1m 15s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe D:\temp\hudson2359985909563698607.sh
+ set +e
+ set +x
INFO: End of cleanup
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

calavera commented:
Given @cpuguy83's heart-y eyes, I'm going to assume that we should move forward with this. I've updated the implementation to shell out already. I've also updated the osxkeychain program to handle that. There are still things to do here, like adding tests and docs, but I'll appreciate your reviews.

vdemeester commented:
If I understand correctly : If there is no credentialStore specified, we use the `FileStore` that does what we do now ; otherwise, it uses a remote one if available and use a `FileStore` to store anything other than username and password ; am I right ? :stuck_out_tongue: 

Design LGTM, I like this a lot :heart_eyes_cat:

calavera commented:
@vdemeester you're correct.

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/14842/console):

~~~console

---
ngratulations!  All commits are properly signed with the DCO!

---> Making bundle: validate-default-seccomp (in bundles/1.11.0-dev/validate-default-seccomp)

---> Making bundle: validate-gofmt (in bundles/1.11.0-dev/validate-gofmt)
These files are not properly gofmt'd:
 - cliconfig/credentials/file_store_test.go
 - cliconfig/credentials/native_store_test.go

Please reformat the above files using "gofmt -s -w" and commit the result.

Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-experimental] $ /bin/sh -xe /tmp/hudson7985952691487878181.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs/23633/console):

~~~console

---
ngratulations!  All commits are properly signed with the DCO!

---> Making bundle: validate-default-seccomp (in bundles/1.11.0-dev/validate-default-seccomp)

---> Making bundle: validate-gofmt (in bundles/1.11.0-dev/validate-gofmt)
These files are not properly gofmt'd:
 - cliconfig/credentials/file_store_test.go
 - cliconfig/credentials/native_store_test.go

Please reformat the above files using "gofmt -s -w" and commit the result.

Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs] $ /bin/sh -xe /tmp/hudson3919988084286467107.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/837/console):

~~~console

---
7f1265.
+ ec=125
+ set +x

----------------------------------
ERROR: Failed to build test binary
----------------------------------



-----------------------------------------------
ERROR: Failed with exitcode 125 at Wed Feb 10 17:27:47 CUT 2016.
-----------------------------------------------


INFO: Tidying up at end of run
INFO: Nuking /d/CI
INFO: Zapped successfully
INFO: End of cleanup
INFO: Ended at Wed Feb 10 17:27:48 CUT 2016 (1m 15s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe D:\temp\hudson3261748282232343397.sh
+ set +e
+ set +x
INFO: End of cleanup
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/21605/console):

~~~console

---
aking bundle: binary (in bundles/1.11.0-dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
# github.com/docker/docker/docker
/usr/local/go/pkg/tool/linux_amd64/link: running gcc failed: fork/exec /usr/bin/gcc: cannot allocate memory

---
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build timed out  : False
Logical operation result is FALSE
Skipping script  : cd c:\scripts
sh timeout-post-build.sh

END OF POST BUILD TASK 	: 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson3337141784447938132.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs/23635/console):

~~~console

---
ate-default-seccomp)

---> Making bundle: validate-gofmt (in bundles/1.11.0-dev/validate-gofmt)
Congratulations!  All Go source files are properly formatted.

---> Making bundle: validate-lint (in bundles/1.11.0-dev/validate-lint)
Errors from golint:
cliconfig/credentials/remote_command_mock.go:19:5: error var cmdError should have name of the form errFoo

Please fix the above errors. You can test via "golint" and commit the result.

Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs] $ /bin/sh -xe /tmp/hudson5935330879305709836.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/14844/console):

~~~console

---
ate-default-seccomp)

---> Making bundle: validate-gofmt (in bundles/1.11.0-dev/validate-gofmt)
Congratulations!  All Go source files are properly formatted.

---> Making bundle: validate-lint (in bundles/1.11.0-dev/validate-lint)
Errors from golint:
cliconfig/credentials/remote_command_mock.go:19:5: error var cmdError should have name of the form errFoo

Please fix the above errors. You can test via "golint" and commit the result.

Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-experimental] $ /bin/sh -xe /tmp/hudson8701352636207512046.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs/23638/console):

~~~console

---
9s	coverage: 19.5% of statements
?   	github.com/docker/docker/builder/dockerfile/command	[no test files]
ok  	github.com/docker/docker/builder/dockerfile/parser	0.011s	coverage: 80.9% of statements
?   	github.com/docker/docker/builder/dockerfile/parser/dumper	[no test files]
ok  	github.com/docker/docker/builder/dockerignore	0.003s	coverage: 93.3% of statements
?   	github.com/docker/docker/cli	[no test files]
ok  	github.com/docker/docker/cliconfig	0.007s	coverage: 89.8% of statements
FAIL	github.com/docker/docker/cliconfig/credentials [build failed]
ok  	github.com/docker/docker/container	0.161s	coverage: 11.0% of statements
?   	github.com/docker/docker/contrib/apparmor	[no test files]
?   	github.com/docker/docker/contrib/docker-device-tool	[no test files]
?   	github.com/docker/docker/contrib/httpserver	[no test files]
ok  	github.com/docker/docker/daemon	0.042s	coverage: 7.5% of statements
ok  	github.com/docker/docker/daemon/events	0.069s	coverage: 39.1% of statements
?   	github.com/docker/docker/daemon/exec	[no test files]
? 
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/14847/console):

~~~console

---
6s	coverage: 19.5% of statements
?   	github.com/docker/docker/builder/dockerfile/command	[no test files]
ok  	github.com/docker/docker/builder/dockerfile/parser	0.009s	coverage: 80.9% of statements
?   	github.com/docker/docker/builder/dockerfile/parser/dumper	[no test files]
ok  	github.com/docker/docker/builder/dockerignore	0.006s	coverage: 93.3% of statements
?   	github.com/docker/docker/cli	[no test files]
ok  	github.com/docker/docker/cliconfig	0.014s	coverage: 89.8% of statements
FAIL	github.com/docker/docker/cliconfig/credentials [build failed]
ok  	github.com/docker/docker/container	0.142s	coverage: 11.0% of statements
?   	github.com/docker/docker/contrib/apparmor	[no test files]
?   	github.com/docker/docker/contrib/docker-device-tool	[no test files]
?   	github.com/docker/docker/contrib/httpserver	[no test files]
ok  	github.com/docker/docker/daemon	0.065s	coverage: 7.5% of statements
ok  	github.com/docker/docker/daemon/events	0.064s	coverage: 39.1% of statements
?   	github.com/docker/docker/daemon/exec	[no test files]
? 
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/844/console):

~~~console

---
vironmentVariablesNodeLoader.gatherEnvironmentVariablesNode(EnvironmentVariablesNodeLoader.java:44)
	... 8 more
Caused by: java.io.IOException
	at hudson.remoting.Channel.close(Channel.java:1163)
	at hudson.slaves.ChannelPinger$1.onDead(ChannelPinger.java:118)
	at hudson.remoting.PingThread.ping(PingThread.java:126)
	at hudson.remoting.PingThread.run(PingThread.java:85)
Caused by: java.util.concurrent.TimeoutException: Ping started at 1455126185831 hasn't completed by 1455126425832
	... 2 more
[PostBuildScript] - Execution post build scripts.
ERROR: Build step failed with exception
java.lang.NullPointerException: no workspace from node hudson.slaves.DumbSlave[azure-windows-tp4-8] which is computer hudson.slaves.SlaveComputer@260708a5 and has channel hudson.remoting.Channel@577078db:azure-windows-tp4-8
	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:74)
	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:64)
	at org.jenkinsci.plugins.postbuildscript.PostBuildScript.processBuildSteps(PostBuildScript.
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/21607/console):

~~~console

---
         125.1 MB
busybox             latest              d9551b4026f0        9 weeks ago         1.113 MB
hello-world         frozen              690ed74de00f        3 months ago        960 B
emptyfs             latest              11f64303f0f7        2 years ago         0 B


-----------------------------------------------
Windows CI Timeout Detection - End of debugging
-----------------------------------------------


c:\scripts>exit 0 
POST BUILD TASK : SUCCESS
END OF POST BUILD TASK : 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson7617607127964923716.sh
+ set +x


---

~~~

aaronlehmann commented on cliconfig/credentials/native_store_test.go:
```diff
@@ -0,0 +1,170 @@
+// +build test_unit
```
Just curious - what's the reason for this build tag?

calavera replied:
I'm using that same tag to load a mock command for testing, to avoid shelling out during tests. I'm still considering if it's the best way to simulate commands. I'm open to suggestions

calavera replied:
See:
https://github.com/docker/docker/pull/20107/files#diff-734ed399478ed3c45691f17639d62e46
and
https://github.com/docker/docker/pull/20107/files#diff-ac0f1eddc56b3465724336aa5a1a649a

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/937/console):

~~~console

---
	      56	docker-1.11.0-dev.exe.md5
  0%  
100%  
	    New File  		      88	docker-1.11.0-dev.exe.sha256
  0%  
100%  
	    New File  		  27.2 m	docker.exe
  0.0%
  3.6%
  7.3%
 11.0%
 14.6%
 18.3%
 22.0%
 25.7%
 29.3%
 33.0%
 36.7%
 40.4%
 44.0%
 47.7%
 51.4%
 55.0%
 58.7%
 62.4%
 66.1%
 69.7%
 73.4%
 77.1%
 80.8%
 84.4%
 88.1%
 91.8%
 95.4%
 99.1%
100%  

------------------------------------------------------------------------------

               Total    Copied   Skipped  Mismatch    FAILED    Extras
    Dirs :         1         0         1         0         0         0
   Files :         4         4         0         0         0         0
   Bytes :   54.45 m   54.45 m         0         0         0         0
   Times :   0:00:00   0:00:00                       0:00:00   0:00:00


   Speed :           1142010840 Bytes/sec.
   Speed :            65346.384 MegaBytes/min.
   Ended : Thursday, February 11, 2016 8:07:40 PM

+ ec=0
+ set +x
INFO: Linking the built binary to /d/CI/CI-085
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-gccgo [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-gccgo/1823/console):

~~~console

---
est.go:1041: TestDockerNetworkDisconnectFromHost.pN59_github_com_docker_docker_integration_cli.DockerNetworkSuite	1.033s
PASS: docker_cli_network_unix_test.go:653: TestDockerNetworkDriverOptions.pN59_github_com_docker_docker_integration_cli.DockerNetworkSuite	0.698s
PASS: docker_cli_network_unix_test.go:850: TestDockerNetworkDriverUngracefulRestart.pN59_github_com_docker_docker_integration_cli.DockerNetworkSuite	4.560s

----------------------------------------------------------------------
FAIL: docker_cli_network_unix_test.go:1006: TestDockerNetworkHostModeUngracefulDaemonRestart.pN59_github_com_docker_docker_integration_cli.DockerNetworkSuite

[d77502000] waiting for daemon to start
[d77502000] daemon started
[d77502000] exiting daemon
[d77502000] waiting for daemon to start
[d77502000] daemon started
docker_cli_network_unix_test.go:1028:
    c.Assert(err, checker.IsNil)
... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc208084cc0)} ("exit status 1")

[d77502000] exiting daemon

----------------------------------------------------------------------
PASS: docker_cli_network_unix_test.go:1069: 
---

~~~

aaronlehmann commented:
I really like this PR. Having secure credential storage will be great.

One suggestion is I think it would be better to serialize the `AuthConfig` struct directly into credential stores instead of extracting the password from that struct. The distribution team is working on adding long-lived tokens for 1.11 that will avoid the need to save passwords at all in many cases. If the credential store treats `AuthConfig` structs as opaque, it should end up being forward-compatible with this model, allowing us to store the long-lived tokens in an OS-specific keychain (which would be the best of both worlds).

calavera commented:
> The distribution team is working on adding long-lived tokens for 1.11 that will avoid the need to save passwords at all in many cases

I don't think there is nothing stopping us from making the necessary adjustment when that happens. We only store strings, it doesn't matter if it's a password or a long-lived token.

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/14913/console):

~~~console

---
tulations! No testing.T found.

---> Making bundle: validate-toml (in bundles/1.11.0-dev/validate-toml)
Congratulations!  All toml source files changed here have valid syntax.

---> Making bundle: validate-vet (in bundles/1.11.0-dev/validate-vet)
vet: cliconfig/credentials/default_store_darwin.go: cliconfig/credentials/default_store_darwin.go:15:7: expected '{', found 'IDENT' err (and 1 more errors)
vet: no files checked
exit status 1
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-experimental] $ /bin/sh -xe /tmp/hudson7336985937746429452.sh
+ set +x


---

~~~

calavera commented:
We should probably consider moving https://github.com/calavera/docker-credential-helpers to the docker organization, anyone against?

GordonTheTurtle commented:
Job: Docker-PRs [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs/23701/console):

~~~console

---
tulations! No testing.T found.

---> Making bundle: validate-toml (in bundles/1.11.0-dev/validate-toml)
Congratulations!  All toml source files changed here have valid syntax.

---> Making bundle: validate-vet (in bundles/1.11.0-dev/validate-vet)
vet: cliconfig/credentials/default_store_darwin.go: cliconfig/credentials/default_store_darwin.go:15:7: expected '{', found 'IDENT' err (and 1 more errors)
vet: no files checked
exit status 1
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs] $ /bin/sh -xe /tmp/hudson3636362011026343216.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-vendor [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-vendor/173/console):

~~~console

---
sks-keyservers.net
gpg: keyserver timed out
gpg: keyserver receive failed: keyserver error
Removing intermediate container 94ae339a7f40
The command '/bin/sh -c apt-key adv --keyserver hkp://p80.pool.sks-keyservers.net:80 --recv-keys E871F18B51E0147C77796AC81196BA81F6B0FC61' returned a non-zero code: 2
Unable to find image 'dockerven:748168d' locally
Pulling repository docker.io/library/dockerven
Error: image library/dockerven:748168d not found
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-vendor] $ /bin/sh -xe /tmp/hudson785553899291925256.sh
+ set +x


---

~~~

thaJeztah commented:
sgtm; perhaps it should be renamed to just 'credential-helpers' ?

calavera commented:
@thaJeztah `engine-credential-helpers`? :wink:

GordonTheTurtle commented:
Job: Docker-PRs-userns [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-userns/6149/console):

~~~console

---
pg: keyserver timed out
gpg: keyserver receive failed: keyserver error
Removing intermediate container 1b3dd82e87ed
The command '/bin/sh -c apt-key adv --keyserver hkp://p80.pool.sks-keyservers.net:80 --recv-keys E871F18B51E0147C77796AC81196BA81F6B0FC61' returned a non-zero code: 2
Unable to find image 'docker:0be9ec6-userns' locally
Pulling repository docker.io/library/docker
Tag 0be9ec6-userns not found in repository docker.io/library/docker
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-userns] $ /bin/sh -xe /tmp/hudson4384997457216629517.sh
+ set +x


---

~~~

vdemeester commented:
SGTM too, but `engine-credential-helpers` seems a little bit too restrictive as any other client (like libcompose) could use it too :angel:. I like `docker-credential-helpers` :stuck_out_tongue_closed_eyes:

calavera commented:
Moved! you should have got some notifications.

GordonTheTurtle commented:
Job: Docker-PRs-gccgo [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-gccgo/1824/console):

~~~console

---
ntegration_cli.DockerSuite	3.353s
PASS: docker_cli_save_load_test.go:76: TestSaveSingleTag.pN52_github_com_docker_docker_integration_cli.DockerSuite	0.546s
PASS: docker_cli_save_load_test.go:22: TestSaveXzAndLoadRepoStdout.pN52_github_com_docker_docker_integration_cli.DockerSuite	3.061s
PASS: docker_cli_save_load_test.go:49: TestSaveXzGzAndLoadRepoStdout.pN52_github_com_docker_docker_integration_cli.DockerSuite	3.153s

----------------------------------------------------------------------
FAIL: docker_cli_search_test.go:28: TestSearchCmdOptions.pN52_github_com_docker_docker_integration_cli.DockerSuite

/go/src/github.com/docker/docker/pkg/integration/dockerCmd_utils.go:35:
    c.Assert(err, check.IsNil, check.Commentf("%q failed with errors: %s, %v", strings.Join(args, " "), out, err))
... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc20848fb00)} ("exit status 1")
... "search -s=2 busybox" failed with errors: Error response from daemon: Get https://index.docker.io/v1/search?q=busybox: dial tcp 52.0.10.162:443: i/o timeout
, exit status 1


---------------
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-gccgo [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-gccgo/1827/console):

~~~console

---
docker_integration_cli.DockerSuite	6.473s
PASS: docker_cli_run_test.go:2701: TestRunRestartMaxRetries.pN52_github_com_docker_docker_integration_cli.DockerSuite	3.706s
PASS: docker_cli_run_test.go:2147: TestRunReuseBindVolumeThatIsSymlink.pN52_github_com_docker_docker_integration_cli.DockerSuite	1.945s
PASS: docker_cli_run_test.go:1179: TestRunRootWorkdir.pN52_github_com_docker_docker_integration_cli.DockerSuite	0.934s

----------------------------------------------------------------------
FAIL: docker_cli_run_unix_test.go:842: TestRunSeccompAllowAptKey.pN52_github_com_docker_docker_integration_cli.DockerSuite

docker_cli_run_unix_test.go:848:
    c.Fatalf("expected apt-key with seccomp to succeed, got %s: %v", out, err)
... Error: expected apt-key with seccomp to succeed, got Executing: gpg --ignore-time-conflict --no-options --no-default-keyring --homedir /tmp/tmp.Rt0I11SM89 --no-auto-check-trustdb --trust-model always --primary-keyring /etc/apt/trusted.gpg --keyring /etc/apt/trusted.gpg.d/debian-archive-jessie-automatic.gpg --keyring /etc/apt/trusted.gpg.d/debian-archive-jessie-security-automat
---

~~~

stevvooe commented on cliconfig/credentials/native_store_test.go:
```diff
@@ -0,0 +1,264 @@
+	inS := string(in)
+
+	switch m.arg {
+	case "erase":
```
:bike: :house: : `delete` or `rm` to be consistent with docker's cli.

calavera replied:
There is a reason I'm using this naming, but I'm open to change it:

```
calavera@d ~ $ docker-credential-osxkeychain
Usage: docker-credential-osxkeychain <store|get|erase>
calavera@d ~ $ git-credential-osxkeychain
usage: git credential-osxkeychain <get|store|erase>
```

:grin:

stevvooe replied:
Interesting. I'll leave this alone if there is an already existing protocol. May be worth noting this somewhere.

calavera replied:
I added a section in the docs about the protocol.

stevvooe commented:
@calavera LGTM on the design. I really like this. Great job! :clap: :dolphin:

vdemeester commented:
I think we can move this to `code-review` :stuck_out_tongue_closed_eyes:

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/1002/console):

~~~console

---
	      56	docker-1.11.0-dev.exe.md5
  0%  
100%  
	    New File  		      88	docker-1.11.0-dev.exe.sha256
  0%  
100%  
	    New File  		  27.2 m	docker.exe
  0.0%
  3.6%
  7.3%
 11.0%
 14.6%
 18.3%
 22.0%
 25.7%
 29.3%
 33.0%
 36.7%
 40.4%
 44.0%
 47.7%
 51.4%
 55.1%
 58.7%
 62.4%
 66.1%
 69.7%
 73.4%
 77.1%
 80.8%
 84.4%
 88.1%
 91.8%
 95.5%
 99.1%
100%  

------------------------------------------------------------------------------

               Total    Copied   Skipped  Mismatch    FAILED    Extras
    Dirs :         1         0         1         0         0         0
   Files :         4         4         0         0         0         0
   Bytes :   54.44 m   54.44 m         0         0         0         0
   Times :   0:00:00   0:00:00                       0:00:00   0:00:00


   Speed :           839556823 Bytes/sec.
   Speed :           48039.826 MegaBytes/min.
   Ended : Friday, February 12, 2016 9:53:51 PM

+ ec=0
+ set +x
INFO: Linking the built binary to /d/CI/CI-5edf66f
---

~~~

calavera commented:
I added some documentation, but I have the feeling that's not enough. I'm open to suggestions.

cpuguy83 commented:
A couple of problems:

1. After aa successful login with the keychain store I cannot pull private images
2. If I've successfully logged in and have an entry for a particular registry in my keychain, if I then remove the auth info from the json config (where it stores my email address for that registry) I get an error about a bad email format. This is fixed by also removing the entry from the keychain.

calavera commented:
> After aa successful login with the keychain store I cannot pull private images

looking

> If I've successfully logged in and have an entry for a particular registry in my keychain, if I then remove the auth info from the json config (where it stores my email address for that registry) I get an error about a bad email format. This is fixed by also removing the entry from the keychain.

that's also the case right now if you remove the email. That field should go away at some point anyways.

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/1077/console):

~~~console

---
	      56	docker-1.11.0-dev.exe.md5
  0%  
100%  
	    New File  		      88	docker-1.11.0-dev.exe.sha256
  0%  
100%  
	    New File  		  27.1 m	docker.exe
  0.0%
  3.6%
  7.3%
 11.0%
 14.7%
 18.4%
 22.0%
 25.7%
 29.4%
 33.1%
 36.8%
 40.5%
 44.1%
 47.8%
 51.5%
 55.2%
 58.9%
 62.5%
 66.2%
 69.9%
 73.6%
 77.3%
 81.0%
 84.6%
 88.3%
 92.0%
 95.7%
 99.4%
100%  

------------------------------------------------------------------------------

               Total    Copied   Skipped  Mismatch    FAILED    Extras
    Dirs :         1         0         1         0         0         0
   Files :         4         4         0         0         0         0
   Bytes :   54.31 m   54.31 m         0         0         0         0
   Times :   0:00:00   0:00:00                       0:00:00   0:00:00


   Speed :           1139142480 Bytes/sec.
   Speed :            65182.255 MegaBytes/min.
   Ended : Tuesday, February 16, 2016 11:38:16 PM

+ ec=0
+ set +x
INFO: Linking the built binary to /d/CI/CI-883
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-vendor [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-vendor/192/console):

~~~console

---
ndor/src/gopkg.in/fsnotify.v1/circle.yml
 D vendor/src/gopkg.in/fsnotify.v1/fsnotify.go
 D vendor/src/gopkg.in/fsnotify.v1/inotify.go
 D vendor/src/gopkg.in/fsnotify.v1/inotify_poller.go
 D vendor/src/gopkg.in/fsnotify.v1/kqueue.go
 D vendor/src/gopkg.in/fsnotify.v1/open_mode_bsd.go
 D vendor/src/gopkg.in/fsnotify.v1/open_mode_darwin.go
 D vendor/src/gopkg.in/fsnotify.v1/windows.go

Please vendor your package with ./hack/vendor.sh.

Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-vendor] $ /bin/sh -xe /tmp/hudson7507156211958954932.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/1124/console):

~~~console

---
	      56	docker-1.11.0-dev.exe.md5
  0%  
100%  
	    New File  		      88	docker-1.11.0-dev.exe.sha256
  0%  
100%  
	    New File  		  27.2 m	docker.exe
  0.0%
  3.6%
  7.3%
 11.0%
 14.7%
 18.3%
 22.0%
 25.7%
 29.4%
 33.0%
 36.7%
 40.4%
 44.1%
 47.7%
 51.4%
 55.1%
 58.8%
 62.4%
 66.1%
 69.8%
 73.5%
 77.2%
 80.8%
 84.5%
 88.2%
 91.9%
 95.5%
 99.2%
100%  

------------------------------------------------------------------------------

               Total    Copied   Skipped  Mismatch    FAILED    Extras
    Dirs :         1         0         1         0         0         0
   Files :         4         4         0         0         0         0
   Bytes :   54.40 m   54.40 m         0         0         0         0
   Times :   0:00:00   0:00:00                       0:00:00   0:00:00


   Speed :           1096992307 Bytes/sec.
   Speed :            62770.403 MegaBytes/min.
   Ended : Wednesday, February 17, 2016 7:13:35 PM

+ ec=0
+ set +x
INFO: Linking the built binary to /d/CI/CI-a2
---

~~~

calavera commented:
@cpuguy83 the pull/push issue should be solved now. I added a test to verify that it works as expected. I also added the most insecure credential helper to the suite, so we can add further tests if we need to:

https://github.com/docker/docker/pull/20107/files#diff-727d759fb79708faee701774f0d1596a

cpuguy83 commented:
LGTM

calavera commented:
@vdemeester, @MHBauer, @runcom 
can I have some :eyes: here, so we can move it to docs review?

vdemeester commented:
LGTM 🐸

thaJeztah commented on docs/reference/commandline/login.md:
```diff
@@ -38,3 +38,74 @@ credentials.  When you log in, the command stores encoded credentials in
+
+## Credentials store
+
+The docker engine can keep user credentials in your native OS keychain.
```
should we limit this description to the keychain, or more generic? perhaps

```Markdown
The Docker Engine can keep user credentials in an external credentials store,
such as the native keychain of the operating system. Using an external store
is more secure than storing credentials in the Docker configuration file.

To use a credentials store, you need an external helper program to interact
with a specific keychain or external store. Docker requires the helper
program to be in the client's host `$PATH`.

This is the list of currently available credentials helpers and where
you can download them from:
```

thaJeztah commented on docs/reference/commandline/login.md:
````diff
@@ -38,3 +38,74 @@ credentials.  When you log in, the command stores encoded credentials in
+```
+
+If you already have credentials set in that file, the best way to move them
+to the native store is to remove them from the file and run `docker login` again.
````
Maybe use `docker logout` to remove the credentials?

```Markdown
If you are currently logged in, run `docker logout` to remove
the credentials from the file and run `docker login` again.
```

thaJeztah commented on docs/reference/commandline/login.md:
```diff
@@ -38,3 +38,74 @@ credentials.  When you log in, the command stores encoded credentials in
+the server address that the docker engine wants to remove credentials for. This is
+an example of that payload: `https://index.docker.io/v1`.
+
+The `erase` command can write error messages in the standard output that the docker engine
```
"in the standard output" -> "to `STDOUT` ? (I think we use that in most places)

thaJeztah commented on docs/reference/commandline/login.md:
```diff
@@ -38,3 +38,74 @@ credentials.  When you log in, the command stores encoded credentials in
+an example of that payload: `https://index.docker.io/v1`.
+
+The `erase` command can write error messages in the standard output that the docker engine
+will show if there was any issue.
```
any issue -> an issue.

thaJeztah commented:
Thanks @calavera some suggestions, but overall quite nice!

calavera commented:
@thaJeztah I addressed your comments, thanks for the suggestions.

thaJeztah commented on docs/reference/commandline/login.md:
```diff
@@ -38,3 +38,77 @@ credentials.  When you log in, the command stores encoded credentials in
+you can download them from:
+
+- Apple OS X keychain: https://github.com/docker/docker-credential-helpers/releases
+- Microsoft Windows Credential Manager: https://github.com/docker/docker-credential-helpers/releases
```
nice! Oh I noticed there's no release yet? https://github.com/docker/docker-credential-helpers/releases

calavera replied:
I'm working on the first release as we speak, but anyone can build the binaries form the source until then :smile_cat:

thaJeztah replied:
Should we link to the repo itself then, so that they end up at the readme?

calavera replied:
no need. We probably have about 1 month to make sure we have releases in the right place.

calavera replied:
:confetti_ball: https://github.com/docker/docker-credential-helpers/releases/tag/v0.1.0 :confetti_ball:

thaJeztah replied:
Yay, congrats! :tada:

thaJeztah commented:
LGTM, thanks!

vdemeester commented:
LGTM 🐯

thaJeztah commented:
All green! Let's merge!

cpuguy83 commented:
:banana:

calavera commented:
:green_apple:

vdemeester commented:
:lemon:

thaJeztah commented:
![image](https://cloud.githubusercontent.com/assets/1804568/13409779/bcdf49c4-df34-11e5-8054-0fb097757a01.jpeg)

JonathonReinhart commented:
I'm not intimately familiar with `git-credential-helper`, but isn't this pretty similar in function? Are existing Git credential helpers compatible with this implementation? If not, is there any reason they had to be different? It seems like a lot of re-inventing the wheel if existing Git credential helpers couldn't be re-used for Docker.

Signed,
A guy running Linux who happily uses Git credential helpers.

jlhawn commented:
@JonathonReinhart I'm also not very familiar with `*-credential-helper`-type applications but I think that the reason why you **should not** reuse an existing credential helper for another application (like `docker` using `git-credential-helper`) is because the the system credential management system typically has access control for which applications are able to access different stored credentials. You certainly *could* do this, but if `docker` used `git-credential-helper` then it would also have access to all of the credentials that `git` would have access to and vice-versa.

By having a separate binary, upon first use, your credential management system will present a dialog asking if you want to authorize this new application to set/access this new entry in your credential store. Now `git-credential-store` will only be able to access your password for `github.com` and not `index.docker.io`, for example.

> A guy running Linux ...

Hmm, looks like just Windows and OS X support for now? I'm sure support for GNOME Keyring will come soon?

runcom commented:
> A guy running Linux ...
> Hmm, looks like just Windows and OS X support for now? I'm sure support for GNOME Keyring will come soon?

working on it https://github.com/docker/docker-credential-helpers/issues/5
